base_layer/mmr/src/sparse_merkle_tree/proofs.rs

```diff
@@ -121,26 +249,41 @@ mod test {
     #[test]
     fn root_proof() {
         let key = NodeKey::from([64u8; 32]);
+        let key2 = NodeKey::from([65u8; 32]);
         let value = ValueHash::from([128u8; 32]);
         let mut tree = SparseMerkleTree::<Blake256>::default();
         let hash = tree.hash().clone();
-        let proof = tree.build_proof(&key).unwrap();
-
-        assert!(!proof.validate_inclusion_proof(&key, &value, &hash));
-        assert!(proof.validate_exclusion_proof(&key, &hash));
+        let in_proof = InclusionProof::from_tree(&tree, &key, &value);
+        assert!(matches!(in_proof, Err(SMTError::NonViableProof)));
+        let ex_proof = ExclusionProof::from_tree(&tree, &key).unwrap();
+        assert!(ex_proof.validate(&key, &hash));
 
         tree.upsert(key.clone(), value.clone()).unwrap();
-        let hash = tree.hash().clone();
-        let proof = tree.build_proof(&key).unwrap();
+        let hash2 = tree.hash().clone();
+
+        let in_proof = InclusionProof::from_tree(&tree, &key, &value).unwrap();
+        let ex_proof = ExclusionProof::from_tree(&tree, &key);
+        assert!(matches!(ex_proof, Err(SMTError::NonViableProof)));
+
+        assert!(in_proof.validate(&key, &value, &hash2));
+        // correct key, wrong value
+        assert!(!in_proof.validate(&key, &ValueHash::from([1u8; 32]), &hash2),);
+        // incorrect key, correct value
+        assert!(!in_proof.validate(&key2, &value, &hash2));
+        // correct key, wrong hash
+        assert!(!in_proof.validate(&key, &value, &hash));
 
-        assert!(proof.validate_inclusion_proof(&key, &value, &hash));
-        assert!(!proof.validate_inclusion_proof(&key, &ValueHash::from([1u8; 32]), &hash),);
-        assert!(!proof.validate_exclusion_proof(&key, &hash));
+        // exclusion proof assertions
+        let ex_proof = ExclusionProof::from_tree(&tree, &key2).unwrap();
+        assert!(!ex_proof.validate(&key, &hash2));
+        assert!(!ex_proof.validate(&key, &hash));
+        assert!(!ex_proof.validate(&key2, &hash));
+        assert!(ex_proof.validate(&key2, &hash2));
     }
 
     #[test]
     fn merkle_proofs() {
```

Comment:
I assume the PRNG seeds used for exclusion proofs happen to exercise both of the disjunctive verification criteria  (empty leaf node and non-matching leaf node), but it isn't clear from just reading the test that this in fact happens. Would recommend a modified or additional test that specifically checks for both cases.